### PR TITLE
#270 Fix application POST stream consumption

### PR DESCRIPTION
**Pull request by multiarc**
Alternative to https://github.com/aspnet/KestrelHttpServer/pull/275

### CODE REVIEW

File: src/Microsoft.AspNet.Server.Kestrel/Http/MessageBody.cs [hunk 1]
```diff
@@ -326,6 +349,17 @@ private static bool TakeChunkedLine(SocketInput baton, ref int chunkSizeOut)
                 }
             }
 
+            protected override void Dispose(bool disposing)
+            {
+                if (_inputLength > 0)
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/MessageBody.cs]: I think this only consumes the next chunk, not the entire request body.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs [hunk 2]
```diff
@@ -137,6 +137,28 @@ public int Peek()
             }
         }
 
+        public MemoryPoolIterator2 Skip(int length)
+        {
+            var block = _block;
+            var index = _index;
+            var remaining = length;
+            var following = block.End - index;
```
Comment (troydai) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: The `following` should be recalculate when moves to a new block.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs [hunk 3]
```diff
@@ -137,6 +137,28 @@ public int Peek()
...
+                {
+                    return new MemoryPoolIterator2(block, index + remaining);
+                }
+                if (block.Next == null)
+                {
+                    return new MemoryPoolIterator2(block, index + following);
+                }
+                remaining -= following;
+                block = block.Next;
+                index = block.Start;
```
Comment (troydai) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: The three branches of condition will be more readable if organized in the form of `if - else if - else`

File: src/Microsoft.AspNet.Server.Kestrel/Http/MessageBody.cs (hunk 1 above)
Comment (multiarc) [inline src/Microsoft.AspNet.Server.Kestrel/Http/MessageBody.cs]: What if we have keep alive connection? Would this run several times for every chunk?

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs (hunk 2 above)
Comment (multiarc) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: Agree, let me fix.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs (hunk 3 above)
Comment (multiarc) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: Not sure how much this would be readable, but added else statements.

**Comment by halter73** [comment]
Closing this seen we merged https://github.com/aspnet/KestrelHttpServer/pull/300 and this doesn't consume all chunks.

If there is a measurable performance improvement, we can reconsider this approach.

**Pull request closed without merging**